File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n;
    static int[] data;
    static int cnt;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        data = new int[n];  // idx에 열, 값에 행을 넣어줌
        DFS(0);
        System.out.println(cnt);
    }

    public static void DFS(int L) {
        if (L == n) { // 마지막까지 도달 -> 종료조건
            cnt += 1;
        } else {
            for (int i = 0; i < n; i++) {  // 가능한 위치 찾기
                data[L] = i;  // 일단 현재 열에 값(행) 넣어두고. 즉 (행(i), 열(L)) 좌표가 가능하다고 해놓고 판단 진행
                if (isValid(L)) {
                    DFS(L + 1);
                }
            }
        }
    }

    public static boolean isValid(int L) {  // 현재 열의 값을 받아옴
        for (int i = 0; i < L; i++) { // 이전 열들까지 (값)행이 같은지
            if (data[L] == data[i]) { // 행이 같은 행이면 놓을 수 없음
                return false;
            } else if (Math.abs(L - i) == Math.abs(data[L] - data[i])) {  // 대각선 판단. 열의 차 와 행의 차가 같으면 같은 대각선
                return false;
            }
        }
        return true;
    }
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys
input = sys.stdin.readline
n = int(input())

data = [0] * n
cnt = 0
visited = [False] * n
def DFS(L):
    global cnt
    if L == n: # 끝까지 도달-> 종료조건
        cnt += 1
    else: # 현재 열에 공격을 받지 않을 행에 퀸을 놓어야 함
        for row in range(n): # 행은 0~n-1에 놓아야 하니까
            if visited[row]:
                continue
            data[L] = row # 현재 열에 row 행에 퀸 위치 (행, 열)
            if isValid(L):
                DFS(L+1)
def isValid(L):
    # 1. 같은 행인지 판단, 2. 같은 대각선인지 판단
    for col in range(L): # 이전 열까지 같은 행이 있는지 판단
        if data[L] == data[col] or abs(L - col) == abs(data[L] - data[col]):
            return False
    return True

DFS(0)
print(cnt)